Implement nested field validators.

markstory opened the pull request:
This adds support for nested field validators. To better support model-less forms and the elasticsearch ODM, we will need to validate nested fields. This set of changes is the smallest change I could make that implements the basic requirements.

Refs #6496

markstory committed:
Implement nested field validators.

This adds support for nested field validators. To better support
model-less forms and the elasticsearch ODM, we will need to validate
nested fields. This set of changes is the smallest change I could make
that implements the basic requirements.

Refs #6496

lorenzo commented on tests/TestCase/Validation/ValidatorTest.php:
```diff
@@ -164,6 +189,33 @@ public function testErrorsWithPresenceRequired()
+    {
+        $validator = new Validator;
+        $validator->add('user.username', 'letter', ['rule' => 'alphanumeric']);
+        $validator->add('comments.0.comment', 'letter', ['rule' => 'alphanumeric']);
```
How do I validate `'comments.[0...N].comment'` ?

markstory replied:
I don't have a good answer for that right now. One option would be to use a limited set of `Hash` path expressions. That would be better than making another path expansion syntax variant.

lorenzo replied:
Or maybe solve this by nesting validators? `$validator->addNested('comments', $anotherValidator)`

markstory replied:
How would that work for the N row case? Would we have separate methods for single objects being nested, and multiples?

lorenzo replied:
Yeah, would that make sense?

markstory replied:
So addNested() and addNestedMultiple()?

I can make thise changes but they will be more invasive than the current changes as ValidationSet will need to work with fields and nested validators.

lorenzo replied:
It doesn't have to be too invasive. In my mind a nested validator was just a custom rule from a custom provider.

markstory replied:
I hadn't thought of it that way. I will update the code tonight.

markstory committed:
Instead of paths, use nested validator objects.

@lorenzo made a great suggestion on how to improve nested validators.
Instead of using paths, we can nest or mount validators inside of each
other. With a few pre-checks this works really well.

markstory commented:
@lorenzo I updated the code to use nested validators. I didn't have a chance to figure out how to pass providers from the root validator into the child validators. I think that could get kind of complicated with validator chains of arbitrary lengths.

lorenzo commented:
This looks great! I'm still thinking if it would actually make sense to copy providers from the parent

lorenzo commented:
Given that no method signature was affected, do we want this in master instead?

markstory commented:
@lorenzo I can copy the providers in when the validator is attached. Should we worry about providers that are added after the nested validator is connected?

lorenzo commented:
@markstory we could maybe make providers an object so that the instance is shared?

markstory commented:
That would fix first level nested validators, but I'm not sure it would fix more deeply nested validators.

markstory committed:
Sync providers from parent validator into children.

Syncing providers allows for a more seamless use of nested validators,
as each nested validator can be assured that it will have the same
rule providers as the other rules.

markstory commented:
@lorenzo I added provider syncing when rules are run this lets us handle late bound providers and deeply nested scenarios.

jhli commented on src/Validation/Validator.php:
```diff
@@ -30,6 +30,12 @@
+     *
+     * @var string
+     */
+    const NESTED = '_nested';
```
I can't grasp this PR but is it made sure that no valdidator called _nested creates an issue? Use __NESTED maybe?

jhli replied:
Or do normal userland validators exclude one starting underscore?

markstory replied:
Generally we try to keep `_` prefixed things as internal use. We do this in the router as well.

jhli replied:
Ok, I hope it is documented in ghe conventions chapter (e.g to avoid underscore prefixed sting identifiers).
Didn't that change for formhelpers options, too? Thanks!

lorenzo commented on src/Validation/Validator.php:
```diff
@@ -195,6 +200,16 @@ public function provider($name, $object = null)
+     */
+    public function providers()
+    {
+        return array_keys($this->_providers);
```
should't this return the whole thing instead?

markstory replied:
I figured we could return the names and let consumers get the providers with provider().

lorenzo commented:
This looks great @markstory, nice job!

What do you think of targeting these changes to master instead?

markstory commented:
@lorenzo I'm ok with re-opening against `master` as no internals changed in the second attempt.

predragleka commented:
Hi guys,

first I want to thank you on a great work you did on 3.x :).

Have a question regarding nested  validation. Is there possibility to get hold of outer context data somehow. In a 'on' callback of inner array fields validation, data inside context is basically sub-document data, and I have need to condition inner data validation based on outer data.

Hope I was clear enough and thanks.

Regards
